### PR TITLE
Closes #237

### LINKED ISSUE #237: Create dataset loader for SciFact
## Adding a Dataset
- **Name:** SciFact
- **Description:** *None provided*
- **Task:** SENT_CLASS|DOC_CLASS
- **Paper:** https://aclanthology.org/2020.emnlp-main.609/
- **Data:** https://github.com/allenai/scifact
- **License:** CC BY-NC 2.0

### PR DESCRIPTION
**Pull request by nbroad1881**
- SciFact
- SciFact is a dataset of 1.4K expert-written scientific claims paired with evidence-containing abstracts, and annotated with labels and rationales.
- https://arxiv.org/pdf/2004.14974.pdf
- https://github.com/allenai/scifact

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [ ] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.

It does not pass the tests because there are multiple configurations for this dataset. bioasq also fails the tests, so I figured this is ok.
#237

### CODE REVIEW

File: biodatasets/scifact/scifact.py [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+from typing import List, Tuple, Dict
+from itertools import chain
+
+import datasets
+from dataclasses import dataclass
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:24]: lets drop this import and instead import the `BigBioConfig` 
`from utils.configs import BigBioConfig`

File: biodatasets/scifact/scifact.py [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+
+@dataclass
+class BigBioConfig(datasets.BuilderConfig):
+    """BuilderConfig for BigBio."""
+
+    name: str = None
+    version: str = None
+    description: str = None
+    schema: str = None
+    subset_id: str = None
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:97]: import this

File: biodatasets/scifact/scifact.py [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+    SOURCE_VERSION = datasets.Version(_SOURCE_VERSION)
+    BIGBIO_VERSION = datasets.Version(_BIGBIO_VERSION)
+
+    BUILDER_CONFIGS = [
+        BigBioConfig(
+            name="scifactcorpus_source",
+            version=SOURCE_VERSION,
+            description="scifact source schema for the corpus config",
+            schema="source",
+            subset_id="scifact",
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:114]: can we use the pattern, 
`name = scifact_corpus_source` 
`subset_id = scifact_corpus` 
for the config names and subset_ids please.

File: biodatasets/scifact/scifact.py [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+        BigBioConfig(
+            name="scifactlabelprediction_bigbio_entailment",
+            version=BIGBIO_VERSION,
+            description="scifact BigBio text entailment schema for label prediction task",
+            schema="bigbio_entailment",
+            subset_id="scifact",
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = "scifact_source"
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:139]: might want to choose one of the source subset_id configs here (e.g. scifact_claims_source)

File: biodatasets/scifact/scifact.py [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+
+    def _info(self) -> datasets.DatasetInfo:
+
+        if self.config.schema == "source":
+            # https://huggingface.co/datasets/scifact/blob/main/scifact.py#L50
+
+            if self.config.name == "scifactcorpus_source":
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:146]:
```suggestion
            if self.config.subset_id == "scifact_corpus":
```

**Review by galtay — CHANGES_REQUESTED** [review]
Thanks for this! With the naming updates I suggested you should be able to run the tests like,
```
python -m tests.test_bigbio biodatasets/scifact/scifact.py --subset_id scifact_corpus
```

**Comment by galtay** [comment]
ps, if you have access to the bioasq dataset locally you can run those tests like, 
```
python -m tests.test_bigbio examples/bioasq.py --subset_id bioasq9b --data_dir ~/data/big_science_biomedical/bioasq
``` 
or after #345 is merged like, 
```
python -m tests.test_bigbio examples/bioasq.py --subset_id bioasq_9b --data_dir ~/data/big_science_biomedical/bioasq
```

File: biodatasets/scifact/scifact.py (hunk 4 above)
Comment (nbroad1881) [reply biodatasets/scifact/scifact.py:139]: Why is it called DEFAULT_CONFIG_NAME when it seems like you want the subset_id to go there?

File: biodatasets/scifact/scifact.py [hunk 6]
```diff
@@ -0,0 +1,428 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Datasets Authors and
+#
+# * Nicholas Broad (nbroad1881)
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:4]: in the old template we were asking people to add names, but eventually decided to just use github to track contributions. can you make this consistent with the current template please https://github.com/bigscience-workshop/biomedical/blob/master/templates/template.py

File: biodatasets/scifact/scifact.py (hunk 4 above)
Comment (nbroad1881) [reply biodatasets/scifact/scifact.py]: > might want to choose one of the source subset_id configs here (e.g. scifact_claims_source)

Did you make a mistake when you said this? 

My understanding is that `DEFAULT_CONFIG_NAME` should be set to a config name (like `scifact_claims_source`) and not to a subset id (like `scifact_claims`). You said subset_id, but then you put a config name

Comment (galtay) [reply biodatasets/scifact/scifact.py]: yes, sorry if this was confusing! i should have said "each config has a unique name, you might want to choose a name from one of the configs implementing the source schema here". each config implementing the source schema has a different subset_id and I dropped too many words between brain and keyboard.

File: biodatasets/scifact/scifact.py [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+                        ),  # Rationale sentences.
+                        "cited_doc_ids": datasets.features.Sequence(
+                            datasets.Value("int32")
+                        ),  # The claim's "cited documents".
+                    }
+                )
+            else:
+                raise NotImplementedError(f"{self.config.subset_id} config not implemented")
+
+        elif self.config.schema == "bigbio_entailment":
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:166]:
```suggestion
        elif self.config.schema == "bigbio_te":
```

File: biodatasets/scifact/scifact.py [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+
+_BIGBIO_ENTAILMENT_LABELPREDICTION_DESCRIPTION = """\
+    SciFact is a dataset of 1.4K expert-written scientific claims paired with evidence-containing abstracts, and annotated with labels and rationales.
+    This task is the following: given a claim and several sentences of evidence (that either support or refute the claim), label that sentence with one of {REFUTES, SUPPORTS, NOINFO}. This corresponds to the third task outlined in Section 5 of the paper.
+    """
+
+_DESCRIPTION = {
+    "scifact_corpus_source": _SOURCE_CORPUS_DESCRIPTION,
+    "scifact_claims_source": _SOURCE_CLAIMS_DESCRIPTION,
+    "scifact_rationale_bigbio_entailment": _BIGBIO_ENTAILMENT_RATIONALE_DESCRIPTION,
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:67]:
```suggestion
    "scifact_rationale_bigbio_te": _BIGBIO_ENTAILMENT_RATIONALE_DESCRIPTION,
```

File: biodatasets/scifact/scifact.py [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+    "scifact_labelprediction_bigbio_entailment": _BIGBIO_ENTAILMENT_LABELPREDICTION_DESCRIPTION,
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:68]:
```suggestion
    "scifact_labelprediction_bigbio_te": _BIGBIO_ENTAILMENT_LABELPREDICTION_DESCRIPTION,
```

File: biodatasets/scifact/scifact.py [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+
+        if "source" in self.config.name:
+            return self._source_generate_examples(filepath, split)
+
+        elif "bigbio" in self.config.name:
+            return self._bigbio_generate_examples(filepath, split)
+
+
+if __name__ == "__main__":
+    datasets.load_dataset(__file__)
```
Comment (galtay) [inline biodatasets/scifact/scifact.py:426]: lets remove this block before merging

**Comment by galtay** [comment]
Just a few small things, but after those changes I think this is ready to merge! 
For any readers, this one doesn't fit well into the testing strategy we had, but all config can be read via, 
```
ds_lab = load_dataset("biodatasets/scifact/scifact.py", name="scifact_labelprediction_bigbio_te")
ds_rat = load_dataset("biodatasets/scifact/scifact.py", name="scifact_rationale_bigbio_te")
ds_corpus = load_dataset("biodatasets/scifact/scifact.py", name="scifact_corpus_source")
ds_claims = load_dataset("biodatasets/scifact/scifact.py", name="scifact_claims_source")
```

**Review by galtay — APPROVED** [review]
:tada:

**Pull request merged**